File: cosky-discovery/src/main/java/me/ahoo/cosky/discovery/ServiceInstanceCodec.java
/*
 * Copyright [2021-2021] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.cosky.discovery;


import lombok.var;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author ahoo wang
 */
public final class ServiceInstanceCodec {
    private static final String SYSTEM_METADATA_PREFIX = "__";
    private static final String METADATA_PREFIX = "_";
    private static final int METADATA_PREFIX_LENGTH = METADATA_PREFIX.length();
    private static final String[] EMPTY_STRING_ARRAY = {};
    private static final String INSTANCE_ID = "instanceId";
    private static final String SERVICE_ID = "serviceId";
    private static final String SCHEMA = "schema";
    private static final String HOST = "host";
    private static final String PORT = "port";
    private static final String WEIGHT = "weight";
    private static final String EPHEMERAL = "ephemeral";
    private static final String TTL_AT = "ttl_at";

    @Deprecated
    public static Map<String, String> encode(ServiceInstance serviceInstance) {
        var serviceInstanceData = new HashMap<String, String>();
        serviceInstanceData.put(INSTANCE_ID, serviceInstance.getInstanceId());
        serviceInstanceData.put(SERVICE_ID, serviceInstance.getServiceId());
        serviceInstanceData.put(SCHEMA, serviceInstance.getSchema());
        serviceInstanceData.put(HOST, serviceInstance.getHost());
        serviceInstanceData.put(PORT, String.valueOf(serviceInstance.getPort()));
        serviceInstanceData.put(WEIGHT, String.valueOf(serviceInstance.getWeight()));
        serviceInstanceData.put(EPHEMERAL, String.valueOf(serviceInstance.isEphemeral()));
        serviceInstance.getMetadata().forEach((key, value) -> {
            var metadataKey = METADATA_PREFIX + key;
            serviceInstanceData.put(metadataKey, value);
        });
        return serviceInstanceData;
    }

    public static String[] encodeMetadata(String[] preArgs, Map<String, String> instanceMetadata) {
        if (instanceMetadata.isEmpty()) {
            return preArgs;
        }
        String[] values = new String[preArgs.length + instanceMetadata.size() * 2];
        System.arraycopy(preArgs, 0, values, 0, preArgs.length);
        var valueIdx = preArgs.length - 1;
        for (Map.Entry<String, String> metadataKV : instanceMetadata.entrySet()) {
            values[++valueIdx] = METADATA_PREFIX + metadataKV.getKey();
            values[++valueIdx] = metadataKV.getValue();
        }
        return values;
    }

    public static ServiceInstance decode(List<String> instanceData) {
        ServiceInstance serviceInstance = new ServiceInstance();
        for (int i = 0; i < instanceData.size(); i = i + 2) {
            var key = instanceData.get(i);
            var value = instanceData.get(i + 1);
            switch (key) {
                case INSTANCE_ID: {
                    serviceInstance.setInstanceId(value);
                    break;
                }
                case SERVICE_ID: {
                    serviceInstance.setServiceId(value);
                    break;
                }
                case SCHEMA: {
                    serviceInstance.setSchema(value);
                    break;
                }
                case HOST: {
                    serviceInstance.setHost(value);
                    break;
                }
                case PORT: {
                    serviceInstance.setPort(Integer.parseInt(value));
                    break;
                }
                case WEIGHT: {
                    serviceInstance.setWeight(Integer.parseInt(value));
                    break;
                }
                case EPHEMERAL: {
                    serviceInstance.setEphemeral(Boolean.parseBoolean(value));
                    break;
                }
                case TTL_AT: {
                    serviceInstance.setTtlAt(Integer.parseInt(value));
                    break;
                }
                default: {
                    if (key.startsWith(SYSTEM_METADATA_PREFIX)) {
                        break;
                    }
                    if (key.startsWith(METADATA_PREFIX)) {
                        var metadataKey = key.substring(METADATA_PREFIX_LENGTH);
                        serviceInstance.getMetadata().put(metadataKey, value);
                    }
                    break;
                }
            }
        }
        return serviceInstance;
    }

    @Deprecated
    public static ServiceInstance decode(Map<String, String> instanceData) {
        ServiceInstance serviceInstance = new ServiceInstance();
        serviceInstance.setInstanceId(instanceData.get(INSTANCE_ID));
        serviceInstance.setServiceId(instanceData.get(SERVICE_ID));
        serviceInstance.setSchema(instanceData.get(SCHEMA));
        serviceInstance.setHost(instanceData.get(HOST));
        serviceInstance.setPort(Integer.parseInt(instanceData.get(PORT)));
        serviceInstance.setWeight(Integer.parseInt(instanceData.get(WEIGHT)));
        serviceInstance.setEphemeral(Boolean.parseBoolean(instanceData.get(EPHEMERAL)));
        if (instanceData.containsKey(TTL_AT)) {
            serviceInstance.setTtlAt(Integer.parseInt(instanceData.get(TTL_AT)));
        }
        instanceData.forEach((key, value) -> {
            if (key.startsWith(METADATA_PREFIX)) {
                var metadataKey = key.substring(METADATA_PREFIX_LENGTH);
                serviceInstance.getMetadata().put(metadataKey, value);
            }
        });
        return serviceInstance;
    }

}

File: cosky-discovery/src/main/kotlin/me/ahoo/cosky/discovery/ServiceInstanceCodec.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.ahoo.cosky.discovery

import me.ahoo.cosky.discovery.ServiceInstance.Companion.TTL_AT_FOREVER

/**
 * Service Instance Codec.
 *
 * @author ahoo wang
 */
object ServiceInstanceCodec {
    private const val SYSTEM_METADATA_PREFIX = "__"
    private const val METADATA_PREFIX = "_"
    private const val METADATA_PREFIX_LENGTH = METADATA_PREFIX.length
    private const val INSTANCE_ID = "instanceId"
    private const val SERVICE_ID = "serviceId"
    private const val SCHEMA = "schema"
    private const val HOST = "host"
    private const val PORT = "port"
    private const val WEIGHT = "weight"
    private const val EPHEMERAL = "ephemeral"
    private const val TTL_AT = "ttl_at"

    fun encodeMetadataKey(key: String): String {
        return METADATA_PREFIX + key
    }

    fun decodeMetadataKey(key: String): String {
        return key.substring(METADATA_PREFIX_LENGTH)
    }

    @JvmStatic
    fun encodeMetadata(preArgs: MutableList<String>, instanceMetadata: Map<String, String>): MutableList<String> {
        if (instanceMetadata.isEmpty()) {
            return preArgs
        }
        instanceMetadata.forEach {
            preArgs.add(encodeMetadataKey(it.key))
            preArgs.add(it.value)
        }
        return preArgs
    }

    fun decode(instanceData: List<String>): ServiceInstance {
        var instanceId: String? = null
        var serviceId: String? = null
        var schema: String? = null
        var host: String? = null
        var port: Int? = null
        var weight: Int? = null
        var isEphemeral: Boolean? = null
        var ttlAt: Long? = null
        val metadata = mutableMapOf<String, String>()
        var i = 0
        while (i < instanceData.size) {
            val key = instanceData[i]
            val value = instanceData[i + 1]
            when (key) {
                INSTANCE_ID -> {
                    instanceId = value
                }

                SERVICE_ID -> {
                    serviceId = value
                }

                SCHEMA -> {
                    schema = value
                }

                HOST -> {
                    host = value
                }

                PORT -> {
                    port = value.toInt()
                }

                WEIGHT -> {
                    weight = value.toInt()
                }

                EPHEMERAL -> {
                    isEphemeral = value.toBoolean()
                }

                TTL_AT -> {
                    ttlAt = value.toLong()
                }

                else -> {
                    if (key.startsWith(METADATA_PREFIX) &&
                        !key.startsWith(SYSTEM_METADATA_PREFIX)
                    ) {
                        val metadataKey = decodeMetadataKey(key)
                        metadata[metadataKey] = value
                    }
                }
            }
            i += 2
        }
        requireNotNull(serviceId) { "serviceId is null" }
        requireNotNull(schema) { "schema is null" }
        requireNotNull(host) { "host is null" }
        requireNotNull(port) { "port is null" }
        requireNotNull(instanceId) { "instanceId is null" }

        val instance = Instance.asInstance(
            serviceId = serviceId,
            schema = schema,
            host = host,
            port = port,
            instanceId = instanceId
        )
        requireNotNull(weight) { "weight is null" }
        requireNotNull(isEphemeral) { "isEphemeral is null" }
        if (isEphemeral) {
            requireNotNull(ttlAt) { "ttlAt is null" }
        }
        return ServiceInstance(
            delegate = instance,
            weight = weight,
            isEphemeral = isEphemeral,
            ttlAt = ttlAt ?: TTL_AT_FOREVER,
            metadata = metadata
        )
    }
}
